[BE-3986] Adds an explicite bodo parser module to calcite

keatoooon opened the pull request:
Associated Bodo PR: https://github.com/Bodo-inc/Bodo/pull/4782

Creates a Bodo Parser module in the calcite repository. Currently, this parser is just a clone of the core parser, but it will likely be extended in future PR's (notably create_table_as).

The Bodo PR just sets the default parser to this new Bodo parser. This should have no issues, since the new Bodo parser is identical to the core parser, but I'm running on nightly to confirm. Please feel free to review this after nightly passes for the above PR.

keatoooon committed:
everything seems to be working, need bodosql tests

keatoooon committed:
resolves github warnings

keatoooon committed:
I hate calcite stype checker

keatoooon committed:
Merge branch 'bodo-calcite-1.30.0-dev' into keat-implement-bodo-parser-module

keatoooon committed:
actually commiting new module now

keatoooon committed:
minor changes

keatoooon committed:
all tests seem to be passing, going to try re-building

keatoooon committed:
commiting needed files to fix thie build

keatoooon committed:
resolved build issues

njriasan commented on .idea/vcs.xml:
```diff
@@ -7,21 +7,4 @@
       <inspection_tool class="SubjectLimit" enabled="true" level="WARNING" enabled_by_default="true" />
     </profile>
   </component>
-  <component name="IssueNavigationConfiguration">
```
What is the reasoning for this change?

keatoooon replied:
I'm not sure what happened here, since this is a generated file. It seems to have reverted itself though.

njriasan commented on .idea/vcs.xml:
```diff
@@ -7,21 +7,4 @@
-    <mapping directory="$PROJECT_DIR$" vcs="Git" />
-  </component>
-</project>
+</project>
```
Add the trailing newline?

njriasan commented on bodo/build.gradle.kts:
```diff
@@ -0,0 +1,72 @@
+
+dependencies {
+    api(project(":core"))
+    api(project(":babel"))
```
If babel is required do we still need to install it directly in the `pom.xml` for BodoSQL? Is it not implicitly captured in the generated JAR file?

Sorry this is not directly related to the changes in this PR but also the Bodo PR.

keatoooon replied:
No, no it is not, I updated the bodo PR.

njriasan commented on bodo/build.gradle.kts:
```diff
@@ -0,0 +1,72 @@
+        it.output.asFileTree.matching { include("**/Parser.jj") }
+    }
+    inputFile.from(parserFile)
+    packageName.set("org.apache.calcite.sql.parser.babel")
```
Is this correct?

keatoooon replied:
Nope fixed

njriasan commented on bodo/build.gradle.kts:
```diff
@@ -0,0 +1,72 @@
+    fun generatedSource(javacc: TaskProvider<org.apache.calcite.buildtools.javacc.JavaCCTask>, sourceSet: String) =
+        generatedJavaSources(javacc.get(), javacc.get().output.get().asFile, sourceSets.named(sourceSet))
+
+    generatedSource(javaCCMain, "main")
```
My gradle knowledge is very limited so I'm not 100% confident this is correct. I flagged one area that seems like it could be wrong but otherwise I'm guessing this is mostly correct.

keatoooon replied:
I just copied this from babel. It seems to work for my IDE, so I'm going to leave it for now.

njriasan commented on bodo/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,575 @@
+    # not a reserved keyword, add it to the 'nonReservedKeywords' section.
+    keywords: [
+      "IF"
+      "SEMI"
```
Is this a hard coded copy?

njriasan replied:
I would remove all of these keywords since we don't support them.

keatoooon replied:
I was just copying directly from babel, I'll clean this up.

njriasan commented on bodo/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,575 @@
+      # are reserved in some version of SQL,
+      # but are not reserved in Babel.
+      #
+      # Words that are commented out (e.g. "AND") are still reserved.
```
Can you add an explanation for structuring things like this or are you just copying directly?

keatoooon replied:
I was just copying directly, I cleaned up all of the config stuff that we're not using.

njriasan commented on bodo/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,575 @@
+    ]
+
+    # List of additional join types. Each is a method with no arguments.
+    # Example: "LeftSemiJoin".
```
I would remove this we don't support semi joins.

njriasan commented on bodo/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,575 @@
+    # Binary operators tokens.
+    # Example: "< INFIX_CAST: \"::\" >".
+    binaryOperatorsTokens: [
+# Temporary reversions for the Bodo version
```
Does this imply we can/should safely revert the original parser once the shift is done?

keatoooon replied:
Yes, but I was going to do that as a follow up since I didn't want to balloon the size of this PR.

njriasan commented on bodo/src/main/codegen/config.fmpp:
```diff
@@ -0,0 +1,575 @@
+    # Return type of method implementation should be "SqlNode".
+    # Example: "DateFunctionCall()".
+    builtinFunctionCallMethods: [
+       "DateFunctionCall()"
```
What are these methods? If this is copied directly perhaps we should remove them and replace it with the empty list.

njriasan commented on bodo/src/main/codegen/includes/parserImpls.ftl:
```diff
@@ -0,0 +1,208 @@
+// limitations under the License.
+-->
+
+JoinType LeftSemiJoin() :
```
Again we don't support SEMI Join so I would consider removing this.

njriasan commented on bodo/src/main/codegen/includes/parserImpls.ftl:
```diff
@@ -0,0 +1,208 @@
+{
+    ( <DATEADD> | <DATEDIFF> | <DATE_PART> ) {
+        s = span();
+        qualifiedName = new SqlIdentifier(unquotedIdentifier(), getPos());
```
Is this to enable passing a time unit in addition to just a string? If so this should be included in the PR title.

keatoooon replied:
No, this was just a copied change, removing it.

njriasan commented on bodo/src/main/codegen/includes/parserImpls.ftl:
```diff
@@ -0,0 +1,208 @@
+    }
+}
+
+boolean IfNotExistsOpt() :
```
It seems to me like a lot of this behavior is copied. I think we should consider limiting this to only those features we intend to support.

njriasan commented on bodo/src/test/java/org/apache/calcite/test/BodoParserTest.java:
```diff
@@ -0,0 +1,361 @@
+ * Tests the Bodo SQL parser. This parser is currently an offshoot of the babel parser,
+ * that allows us to extend the parser to suit our own purposes.
+ *
+ * Currently, this only contains tests equivalent to Babel, but this can and should
```
I would try removing behavior we don't support and shrinking these tests if its not too much work.

njriasan commented on bodo/src/test/resources/sql/big-query.iq:
```diff
@@ -0,0 +1,137 @@
+# big-query.iq - Babel test for BigQuery dialect of SQL
```
Is this necessary?

njriasan commented on bodo/src/test/resources/sql/redshift.iq:
```diff
@@ -0,0 +1,2206 @@
+# redshift.iq - Babel test for Redshift dialect of SQL
```
Is this necessary?

njriasan reviewed (commented):
Overall I think this looks good, but I think we should consider restricting this to just functionality we support right now to limit the number of changes.

keatoooon committed:
Merge branch 'bodo-calcite-1.30.0-dev' into keat-implement-bodo-parser-module

keatoooon committed:
cleanup, still having issue with bodo parser module name

keatoooon committed:
reverting changes to generated ide file

keatoooon committed:
it's working now. I have no idea why, but I'm not complaining

keatoooon committed:
removing last kw

keatoooon committed:
removed unused testing files

keatoooon commented:
Addressed comments, nightly run is here: https://dev.azure.com/bodo-inc/Bodo/_build/results?buildId=22014&view=results

Will mark as ready for review once this passes.

keatoooon committed:
updated IF to be non-reserved

keatoooon commented:
This is a little block rn because azure CI is backloged. I've fixed the errors found on the previous azure CI run, but I'd like to run it once more, to be certain I got them all.

keatoooon commented:
Everything passed except for a few error handling tests (because the default looka head is now slightly larger) and `test_snowflake_catalog_default`. I'm hoping `test_snowflake_catalog_default` is transient, I'm re-running nightly now to be sure: https://dev.azure.com/bodo-inc/Bodo/_build/results?buildId=22065&view=results

keatoooon commented:
Everything passed here: https://dev.azure.com/bodo-inc/Bodo/_build/results?buildId=22065&view=logs&jobId=00b7a64d-2004-54e2-21f4-7be093358810&j=00b7a64d-2004-54e2-21f4-7be093358810&t=d772d261-6072-527c-1182-b4bcdab45578 (unrelated Bodo error)

keatoooon reviewed (commented):
Spoke with Jonathan offline. Feedback was:
1. Remove create to followup PR
2. Remove babel dependency and/or make Bodo versions of the needed babel tests

keatoooon committed:
removed dependency on babel

keatoooon committed:
Merge branch 'bodo-calcite-1.30.0-dev' into keat-implement-bodo-parser-module

keatoooon commented:
Green CI run with the new changes here: https://dev.azure.com/bodo-inc/Bodo/_build/results?buildId=22070&view=results

jsternberg reviewed (approved):
I'm good with this so we can start using the new parser and make iterations and fixes later.

keatoooon committed:
Did some comment cleanup

keatoooon committed:
a bit more cleanup

keatoooon committed:
removed volatile

keatoooon committed:
removed IF

keatoooon committed:
final cleanup

keatoooon committed:
reverting volatile and IF changes, it's not worth the potential issues down the line
